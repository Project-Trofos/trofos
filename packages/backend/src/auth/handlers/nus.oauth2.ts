import axios from "axios"
import jwt from "jsonwebtoken";

const HANDLER_NAME = 'NUS_OAUTH2';
const BASE_URL = 'https://luminus.nus.edu.sg/v2/api/login/adfstoken';
const GRANT_TYPE = "authorization_code";
const CLIENT_ID = "INC000002767827";
const RESOURCE = "sg_edu_nus_oauth";
const TOKEN_EMAIL_ADDRESS_KEY = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"

async function handleLogin(code: string, callbackUrl: string) {
    const data = new URLSearchParams();
    data.append('grant_type', GRANT_TYPE);
    data.append('client_id', CLIENT_ID);
    data.append('resource', RESOURCE);
    data.append('code', code);
    data.append('redirect_uri', callbackUrl);

    const config = {
        method: 'post',
        url: BASE_URL,
        data
      };
      

      const token = await axios(config);
      if (!token.data.access_token) {
        throw new Error("Error while fetching token from server");
      }

      const decodedToken = jwt.decode(token.data.access_token, {json : true});
      if (!decodedToken) {
        throw new Error("Error while decoding access token");
      }

      return decodedToken[TOKEN_EMAIL_ADDRESS_KEY];
}

export default {
    HANDLER_NAME,
    handleLogin
}