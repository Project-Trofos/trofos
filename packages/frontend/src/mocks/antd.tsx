import React from 'react';
import { Select as AntdSelect, DatePicker as AntdDatePicker } from 'antd';
import mockDayjs from 'dayjs';

// Mock certain antd component to simplify testing in React Testing Library
vi.mock('antd', async () => { 
  const antd: any = await vi.importActual('antd');
  antd.theme.defaultConfig.hashed = false;

  function DatePicker(props: React.ComponentProps<typeof AntdDatePicker>) {
    const { defaultValue, placeholder, disabled, id, onChange } = props;
    return (
      <input
        id={id}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        placeholder={placeholder}
        disabled={disabled}
        onChange={(e) => {
          if (onChange) {
            onChange(mockDayjs(e.target.value) as unknown as Parameters<typeof onChange>['0'], e.target.value);
          }
        }}
      />
    );
  }

  function RangePicker(props: React.ComponentProps<typeof AntdDatePicker.RangePicker>) {
    const { defaultValue, id, onChange } = props;
    return (
      <input
        id={id as string | undefined}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        onChange={(e) => {
          if (onChange) {
            onChange(
              [mockDayjs(e.target.value), mockDayjs(e.target.value)] as unknown as Parameters<typeof onChange>['0'],
              [e.target.value, e.target.value],
            );
          }
        }}
      />
    );
  }

  DatePicker.RangePicker = RangePicker;

  function Select(props: React.ComponentProps<typeof AntdSelect>) {
    const { mode, value, defaultValue, className, onChange, disabled, children, id } = props;

    const multiple = ['tags', 'multiple'].includes(mode ?? '');

    return (
      <select
        id={id}
        value={value as React.SelectHTMLAttributes<HTMLSelectElement>['value']}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLSelectElement>['defaultValue']}
        multiple={multiple}
        disabled={disabled}
        className={className}
        onChange={(e) => {
          if (onChange) {
            onChange(
              multiple ? Array.from(e.target.selectedOptions).map((option) => option.value) : e.target.value,
              {},
            );
          }
        }}
      >
        {children}
      </select>
    );
  }

  function Option({ children, ...otherProps }: { children: React.ReactNode }) {
    return <option {...otherProps}>{children}</option>;
  }
  function OptGroup({ children, ...otherProps }: { children: React.ReactNode }) {
    return <optgroup {...otherProps}>{children}</optgroup>;
  }

  Select.Option = Option;
  Select.OptGroup = OptGroup;

  return { ...antd, Select, DatePicker };
});
