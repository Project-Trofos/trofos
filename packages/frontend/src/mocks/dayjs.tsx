const dayjsMock = {
  year: () => '2022',
};

export default dayjsMock;
