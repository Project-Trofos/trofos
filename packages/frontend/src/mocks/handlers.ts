/* eslint-disable import/no-extraneous-dependencies */
import { rest } from 'msw';
import { CourseData } from '../api/types';

const BASE_URL = process.env.REACT_APP_BACKEND_BASE_URL || 'http://localhost:3001';
const NUSMODS_URL = 'https://api.nusmods.com/v2/2022-2023/moduleList.json';

export const MSW_COURSE: CourseData = {
  id: 1,
  code: 'CS3203',
  cname: 'Software Engineering Project',
  startYear: 2022,
  endYear: 2022,
  startSem: 1,
  endSem: 1,
  description: null,
  milestones: [
    {
      id: 1,
      name: 'Milestone 1',
      course_id: 1,
      start_date: '2023-12-31',
      deadline: '2023-01-09',
      created_at: '2022-12-18',
    },
    {
      id: 2,
      name: 'Milestone 2',
      course_id: 1,
      start_date: '2023-01-10',
      deadline: '2023-01-19',
      created_at: '2022-12-18',
    },
    {
      id: 3,
      name: 'Milestone 3',
      course_id: 1,
      start_date: '2023-01-20',
      deadline: '2023-01-30',
      created_at: '2022-12-18',
    },
  ],
  users: [],
  public: false,
  created_at: '2022-09-14T03:33:34.960Z',
};

const PROJECT = {
  id: 1,
  pname: 'project1',
  pkey: null,
  description: 'project1_description',
  course_id: '1',
  public: false,
  created_at: '2022-09-15T01:58:01.735Z',
  course: MSW_COURSE,
};

const NUSMODS_MODULE = {
  moduleCode: 'EL1101E',
  title: 'The Nature of Language',
  semesters: [1, 2],
};

const handlers = [
  // Handles GET on /project
  rest.get(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([PROJECT])))),

  // Handles GET on /course
  rest.get(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([MSW_COURSE])))),

  // Handles GET to NUSMODS
  rest.get(NUSMODS_URL, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([NUSMODS_MODULE])))),

  // Handles POST on /project
  rest.post(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/project`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  rest.post(`${BASE_URL}/course/:courseId/milestone`, (req, res, ctx) =>
    res(ctx.status(200), ctx.body(JSON.stringify({}))),
  ),
];

export default handlers;
