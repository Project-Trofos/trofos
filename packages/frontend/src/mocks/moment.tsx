const momentMock = {
  year: () => '2022',
};

export default momentMock;
