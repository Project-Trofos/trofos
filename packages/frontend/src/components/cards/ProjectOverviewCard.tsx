import React from 'react';
import { Card, Row, Col, Statistic, Divider, Space, Typography } from 'antd';
import Countdown from 'antd/es/statistic/Countdown';
import { Sprint } from '../../api/sprint';
import { Backlog, BacklogHistory } from '../../api/types';
import { Subheading } from '../typography';
import DailyCompletedPointsBarGraph from '../visualization/DailyCompletedPointsBarGraph';
import SprintBacklogPieChart from '../visualization/SprintBacklogPieChart';

export default function ProjectOverviewCard(props: {
  sprints: Sprint[];
  unassignedBacklogs: Backlog[];
  backlogHistory: BacklogHistory[];
}) {
  const { sprints, unassignedBacklogs, backlogHistory } = props;
  const activeSprint = sprints.find((s) => s.status === 'current');
  return (
    <Card style={{ height: '100%' }}>
      {activeSprint ? (
        <>
          <Row gutter={[32, 32]}>
            <Col sm={6} xs={24}>
              <Statistic title="Current Active Sprint" value={activeSprint.name} />
            </Col>
            <Col sm={6} xs={24}>
              <Statistic
                title="Completed Issues"
                value={activeSprint.backlogs.filter((b) => b.status === 'Done').length}
                suffix={`/ ${activeSprint.backlogs.length}`}
              />
            </Col>
            <Col sm={6} xs={24}>
              <Statistic title="Backlogs" value={unassignedBacklogs.length} />
            </Col>
            <Col sm={6} xs={24}>
              <Countdown title="Time before end of sprint" value={activeSprint.end_date} format="D day HH hour" />
            </Col>
          </Row>
          <Divider />
          <Row>
            <Col xs={24} md={12}>
              <Space direction="vertical" style={{ width: '100%', alignItems: 'center' }}>
                <Subheading>Active Sprint Issue Types</Subheading>
                <SprintBacklogPieChart sprint={activeSprint} unassignedBacklog={unassignedBacklogs} />
              </Space>
            </Col>
            <Col xs={24} md={12}>
              <Space direction="vertical" style={{ width: '100%', alignItems: 'center' }}>
                <Subheading>Daily Completed Story Points</Subheading>
                <DailyCompletedPointsBarGraph
                  backlogHistory={backlogHistory.filter((b) => b.sprint_id === activeSprint.id)}
                />
              </Space>
            </Col>
          </Row>
        </>
      ) : (
        <Typography.Text>
          There are no active sprint currently, start a sprint to see statistics related to that sprint.
        </Typography.Text>
      )}
    </Card>
  );
}
