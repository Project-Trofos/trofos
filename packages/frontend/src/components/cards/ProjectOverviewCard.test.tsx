import React from 'react';
import { render, screen } from '@testing-library/react';
import ProjectOverviewCard from './ProjectOverviewCard';
import { MSW_BACKLOG_HISTORY, MSW_SPRINT } from '../../mocks/handlers';

describe('ProjectOverviewCard test', () => {
  describe('when there are no active sprint', () => {
    it('renders card with prompt about no active sprint', () => {
      render(<ProjectOverviewCard sprints={[MSW_SPRINT[0]]} backlogHistory={[]} unassignedBacklogs={[]} />);
      expect(screen.getByText('no active sprint', { exact: false })).toBeInTheDocument();
    });
  });

  describe('when there is active sprint', () => {
    it('renders card correctly', () => {
      const { baseElement } = render(
        <ProjectOverviewCard sprints={MSW_SPRINT} backlogHistory={MSW_BACKLOG_HISTORY} unassignedBacklogs={[]} />,
      );

      expect(baseElement).toMatchSnapshot();
    });
  });
});
