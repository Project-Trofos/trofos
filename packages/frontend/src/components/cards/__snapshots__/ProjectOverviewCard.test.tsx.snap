// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectOverviewCard test when there is active sprint renders card correctly 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered"
      style="height: 100%;"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="ant-row"
          style="margin: -16px -16px -16px -16px;"
        >
          <div
            class="ant-col ant-col-xs-24 ant-col-sm-6"
            style="padding: 16px 16px 16px 16px;"
          >
            <div
              class="ant-statistic"
            >
              <div
                class="ant-statistic-title"
              >
                Current Active Sprint
              </div>
              <div
                class="ant-statistic-content"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  Sprint 2
                </span>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-xs-24 ant-col-sm-6"
            style="padding: 16px 16px 16px 16px;"
          >
            <div
              class="ant-statistic"
            >
              <div
                class="ant-statistic-title"
              >
                Completed Issues
              </div>
              <div
                class="ant-statistic-content"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    2
                  </span>
                </span>
                <span
                  class="ant-statistic-content-suffix"
                >
                  / 3
                </span>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-xs-24 ant-col-sm-6"
            style="padding: 16px 16px 16px 16px;"
          >
            <div
              class="ant-statistic"
            >
              <div
                class="ant-statistic-title"
              >
                Backlogs
              </div>
              <div
                class="ant-statistic-content"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  <span
                    class="ant-statistic-content-value-int"
                  >
                    0
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-xs-24 ant-col-sm-6"
            style="padding: 16px 16px 16px 16px;"
          >
            <div
              class="ant-statistic css-dev-only-do-not-override-sk7ap8"
            >
              <div
                class="ant-statistic-title"
              >
                Time before end of sprint
              </div>
              <div
                class="ant-statistic-content"
              >
                <span
                  class="ant-statistic-content-value"
                >
                  12 day 19 hour
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-divider ant-divider-horizontal"
          role="separator"
        />
        <div
          class="ant-row"
        >
          <div
            class="ant-col ant-col-xs-24 ant-col-md-12"
          >
            <div
              class="ant-space ant-space-vertical"
              style="width: 100%; align-items: center;"
            >
              <div
                class="ant-space-item"
                style="margin-bottom: 8px;"
              >
                <h4
                  class="ant-typography"
                >
                  Active Sprint Issue Types
                </h4>
              </div>
              <div
                class="ant-space-item"
              >
                <div
                  data-chart-source-type="G2Plot"
                  size-sensor-id="1"
                  style="height: inherit;"
                >
                  <div
                    style="position:relative;"
                  >
                    <canvas
                      height="400"
                      style="width: 400px; height: 400px; display: inline-block; vertical-align: middle;"
                      width="400"
                    />
                  </div>
                  <object
                    class="size-sensor-object"
                    data="about:blank"
                    style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
                    tabindex="-1"
                    type="text/html"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-xs-24 ant-col-md-12"
          >
            <div
              class="ant-space ant-space-vertical"
              style="width: 100%; align-items: center;"
            >
              <div
                class="ant-space-item"
                style="margin-bottom: 8px;"
              >
                <h4
                  class="ant-typography"
                >
                  Daily Completed Story Points
                </h4>
              </div>
              <div
                class="ant-space-item"
              >
                <div
                  data-chart-source-type="G2Plot"
                  size-sensor-id="2"
                  style="height: inherit;"
                >
                  <div
                    style="position:relative;"
                  >
                    <canvas
                      height="400"
                      style="width: 400px; height: 400px; display: inline-block; vertical-align: middle;"
                      width="400"
                    />
                  </div>
                  <object
                    class="size-sensor-object"
                    data="about:blank"
                    style="display: block; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1; opacity: 0;"
                    tabindex="-1"
                    type="text/html"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
